cpu/kinetis_common/hwtimer_arch.c

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright (C) 2014 PHYTEC Messtechnik GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     cpu_kinetis_common
+ * @{
+ *
+ * @file
+ * @brief       Implementation of the kernels hwtimer interface
+ *              hwtimer uses Freescale Low Power Timer lptmr0.
+ *              There are two clock sources supported and tested:
+ *              LPO - 1kHz
+ *              RTC - 32768Hz
+ *
+ * @author      Johann Fischer <j.fischer@phytec.de>
+ *
+ * @}
+ */
+
+#include "arch/hwtimer_arch.h"
+#include "hwtimer_cpu.h"
+#include "cpu-conf.h"
+#include "thread.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+#define LPTMR_MAXTICKS            (0x0000FFFF)
+
+#ifndef LPTIMER_CNR_NEEDS_LATCHING
+#warning LPTIMER_CNR_NEEDS_LATCHING is not defined in cpu-conf.h! Defaulting to 1
+#define LPTIMER_CNR_NEEDS_LATCHING 1
+#endif
+
+typedef struct {
+    uint32_t counter32b;
+    uint32_t cmr32b;
+    uint32_t diff;
+} hwtimer_stimer32b_t;
+
+static hwtimer_stimer32b_t stimer;
+
+/**
+ * @brief Reference to the hwtimer callback
+ */
+void (*timeout_handler)(int);
+
+inline static uint32_t lptmr_get_cnr(void)
+{
+#if LPTIMER_CNR_NEEDS_LATCHING
+    /* Write some garbage to CNR to latch the current value */
+    LPTMR0->CNR = 42;
+    return (uint32_t)LPTMR0->CNR;
+#else
+    /* The early revisions of the Kinetis CPUs do not need latching of the CNR
+     * register. However, this may lead to corrupt values, we read it twice to
+     * ensure that we got a valid value */
+    int i;
+    uint32_t tmp;
+    uint32_t cnr = LPTMR0->CNR;
+
+    /* you get three retries */
+    for (i = 0; i < 3; ++i) {
+        tmp = LPTMR0->CNR;
+
+        if (tmp == cnr) {
+            return cnr;
+        }
+
+        cnr = tmp;
+    }
+
+#endif
+}
+
+inline static void hwtimer_start(void)
+{
+    LPTIMER_DEV->CSR |= LPTMR_CSR_TEN_MASK;
+}
+
+inline static void hwtimer_stop(void)
+{
+    LPTIMER_DEV->CSR &= ~LPTMR_CSR_TEN_MASK;
+}
+
+void hwtimer_arch_init(void (*handler)(int), uint32_t fcpu)
+{
+    timeout_handler = handler;
+
+    /* unlock LPTMR0 */
+    LPTIMER_CLKEN();
+    /* set lptmr's IRQ priority */
+    NVIC_SetPriority(LPTIMER_IRQ_CHAN, LPTIMER_IRQ_PRIO);
+    /* reset lptmr */
+    LPTIMER_DEV->CSR = 0;
+
+    switch (LPTIMER_CLKSRC) {
+        case LPTIMER_CLKSRC_MCGIRCLK:
+            /* Select MCGIRCLK as clock source */
```

Comment:
But that is a option for MCG driver or your cpu_clock_init, not for LPTMR driver.